debian/liesel.1
.\" Manpage for liesel
.\" Contact andrew@rail5.org to correct errors or typos.
.TH man 8 "23 January 2022" "10.0" "liesel man page"
.SH NAME
liesel \- Transform ordinary PDFs into booklets
.SH SYNOPSIS
liesel -i some-book.pdf -o ready-to-print.pdf
.SH DESCRIPTION
Liesel prepares an inputted PDF to be DIY printed at home in pamphlet- or booklet-style
.SH OPTIONS
  -i
  --input
    PDF to convert

  -o
  --output
    New file for converted PDF

  -I
  --stdin
    Read PDF from STDIN rather than file
    e.g: liesel -I -o output.pdf < input.pdf

  -O
  --stdout
    Write output PDF to STDOUT rather than file
    e.g: liesel -i input.pdf -O > output.pdf

  -e
  --export
    Export example/preview JPEG image of selected pages
    e.g: -i infile.pdf -e 5,6 -o preview.jpeg
    e.g: --input infile.pdf --export 5,6 --output preview.jpeg

  -g
  --greyscale
  --grayscale
    Convert PDF to greyscale/black and white

  -r
  --range
    Print only specified range of pages (in the order supplied)
    e.g: -r 1-12
    e.g: --range 15-20,25-30
    e.g: -r 10,9,5,2,1
    e.g: --range 20-10
    
    The -r option will process only the selection you give, and in the order that you give it
    "-r 1,2,3,4,5" is the same as "-r 1-5"
    "-r 5-1" will print those pages in descending order

    You can also combine ranges using commas:
    -r 1-5,20-25,7,5-1
    will print first 1,2,3,4,5
    then 20,21,22,23,24,25
    then 7
    then 5,4,3,2,1

  -s
  --segment
    Print output PDFs in segments of a given size
    e.g: -s 40
      (produces multiple PDFs of 40 pages each)

  -m
  --minsize
    Specify minimum segment size (default is 4)
    e.g: -m 8
    
    The minimum segment size is only used when printing in segments
    If, for example, you're printing 100 pages with -s 40
    then 2 segments of 40 + 1 "runoff" segment of 20 will be produced
    If, instead, you print those 100 pages with -s 40 -m 21
    then those 20 pages, since they don't meet the minimum segment size,
    will be tacked on to the previous segment,
    making 1 segment of 40 pages + 1 segment of 60

    If -m is not specified, it defaults to 4 pages

  -f
  --force
    Force overwrites
      (do not warn about files already existing)

  -v
  --verbose
    Verbose mode

  -b
  --bookthief
  --progress
    Show progress (percentage done)

  -d
  --density
  --quality
  --ppi
    Specify pixels-per-inch density/quality (default is 100)
    e.g: -d 200

  -t
  --transform
  --rescale
    Transform output PDF to print on a specific size paper
      e.g: -t 8.5x11
      or: -t 8.5x11
      or: -t 23x7
      etc

  -l
  --landscape
  --duplex
    Duplex printer "landscape" flipping compatibility
      (flips every other page)

  -k
  --threshold
    Convert to pure black-and-white (not grayscale) with given threshold value between 0-100
    e.g: -k 70
      will convert any color with brightness under 70% to pure black, and any brighter color to pure white
      this option is particularly useful in printing PDFs of scanned books with yellowed pages etc

  -C
  --crop
    Crop pages according to specified values
    e.g: -C 10,20,30,40
      order: left,right,top,bottom
      cuts 10% from the left side, 20% from the right side, etc

  -w
  --widen
    Widen center margins according to specified value
    e.g: -w 30
      (adds blank space between pages)

  -a
  --auto-widen
    Auto-widen center margins
    e.g: -a (completely automatic)
    e.g: -a 30 (auto, with a maximum of 30)
      (progressively widens center margins towards the middle of the booklet)
      (if -w / --widen is also specified, -w / --widen acts as the minimum margin)

  -D
  --divide
    Divide each page into two
      divides the left half and the right half into separate pages
      this option is particularly useful in printing PDFs of scanned books which haven't already separated the pages
  
  -N
  --no-booklet
  --linear
    Applies the changes requested (crop, color threshold, etc) and outputs linearly, without making a booklet
      all features can be used except for center-margin widening

  -p
  --pages
    Count pages of input PDF and exit

  -c
  --check
    Check validity of command, and do not execute
  
  -B
  --output-settings
    Prints the parameters/settings you've provided, in XML format, to STDOUT

  -h
  --help
    Print this help message

  -q
  --info
    Print program info

  -V
  --version
    Print version number

Example:
  liesel -i some-book.pdf -g -r 64-77 -f -d 150 -v -b -o ready-to-print.pdf
  liesel -i some-book.pdf -r 100-300,350-400,1-10 -s 40 -t 8.25x11.75 -m 16 -o ready-to-print.pdf
  liesel --input some-book.pdf --range 1,5,7,3,1,50 --landscape --output ready-to-print.pdf
  liesel -p some-book.pdf
  liesel -c -i some-book.pdf -o output.pdf
.SH SEE ALSO
bookthief
.SH BUGS
No known bugs.
.SH AUTHOR
rail5 (andrew@rail5.org)


debian/changelog
liesel (11.3.7) stable; urgency=medium

  * Cleaned up debian packaging

 -- rail5 <andrew@rail5.org>  Tue, 04 Jun 2024 13:16:07 +0800

liesel (11.3.6) stable; urgency=medium

  * Meaningless update to test autobuild distribution. Nothing
    has changed.

 -- rail5 <andrew@rail5.org>  Wed, 01 May 2024 22:02:05 +0800

liesel (11.3.5) stable; urgency=medium

  * Moving everything to Debian 12

 -- rail5 <andrew@rail5.org>  Fri, 18 Aug 2023 23:13:29 -0400

liesel (11.3.4) bullseye; urgency=medium

  * Set libtiff6 and libwebp7 as usable alternatives
  to libtiff5 and libwebp6 so that Liesel can now
  be installed on Debian 12

 -- rail5 <andrew@rail5.org>  Tue, 20 Jun 2023 14:09:47 -0400

liesel (11.3.3) bullseye; urgency=medium

  * Patched a bug which prevented use of PDF files
  larger than 2GB on 32bit systems

 -- rail5 <andrew@rail5.org>  Sun, 02 Apr 2023 14:20:49 -0400

liesel (11.3.2) bullseye; urgency=medium

  * Adding i386 builds to deb.rail5.org

 -- rail5 <andrew@rail5.org>  Fri, 31 Mar 2023 21:38:28 -0400

liesel (11.3.1) focal; urgency=low

  * Liesel can now compile on OpenBSD

 -- rail5 <andrew@rail5.org>  Tue, 14 Feb 2023 01:13:12 -0500

liesel (11.3) focal; urgency=low

  * Liesel can now compile on MacOS

 -- rail5 <andrew@rail5.org>  Tue, 15 Nov 2022 16:55:33 -0500

liesel (11.2) focal; urgency=low

  * Patched a bug in export preview

 -- rail5 <andrew@rail5.org>  Thu, 31 Mar 2022 17:50:13 -0400

liesel (11.1) focal; urgency=low

  * Patched a series of bugs listed on github
  Average execution sped up as a result, using libHaru
  to write the output PDFs. Also as a result, Liesel
  no longer cross-compiles to Windows

 -- rail5 <andrew@rail5.org>  Tue, 8 Mar 2022 01:08:40 -0500

liesel (11.0) focal; urgency=low

  * Added new -N feature, which applies requested changes
  but does not make a booklet
  Dramatically restructured the code to be more object-oriented
  to make for easier changes in the future
  Patched a few bugs along the way

 -- rail5 <andrew@rail5.org>  Sun, 30 Jan 2022 19:32:22 -0500

liesel (10.1) focal; urgency=low

  * Shipping previous change
  Also added a manpage entry for -B

 -- rail5 <andrew@rail5.org>  Fri, 28 Jan 2022 00:53:31 -0500

liesel (10.1~alpha1) focal; urgency=low

  * Added --output-settings longopt for -B
  Added -B entry to help menu
  -B now outputs an XML table

 -- rail5 <andrew@rail5.org>  Thu, 27 Jan 2022 22:06:48 -0500

liesel (10.0) focal; urgency=low

  * v10.0 release
  Added longopts support for easier use
  Added option to receive PDF from STDIN (-I / --stdin)
  And option to output PDF to STDOUT (-O / --stdout)

 -- rail5 <andrew@rail5.org>  Sun, 23 Jan 2022 23:39:12 -0500

liesel (9.0) focal; urgency=low

  * v9.0 release
  Added -a 'auto margin' feature
  progressively widens center-margins
  (the space between the left and right-hand pages)
  towards the middle of the booklet,
  so that those margins are wider in the middle,
  and narrower at the ends
  primarily for use in printing very large booklets
  Added undocumented -B 'bookthief extra' option
  outputs a newline-delimited list of user-provided options
  used for BookThief's new 'import command' feature
  Separated native-nix compilation and windows cross-compilation
  in the makefile

 -- rail5 <andrew@rail5.org>  Sat, 1 Jan 2022 22:01:17 -0500

liesel (8.1) focal; urgency=low

  * Cleaned up the code a bit
  Organized includes/functions
  Attempted to make the code easier to navigate
  Slightly re-structured main() function to remove
  unnecessary conditionals

 -- rail5 <andrew@rail5.org>  Fri, 31 Dec 2021 01:19:42 -0500

liesel (8.0) focal; urgency=low

  * v8.0 release
  Added -k 'color threshold' feature
  converts the PDF into pure black-and-white
  with a given 'threshold' value
  useful for handling scanned PDFs with yellowed pages
  Added -w 'widen center margins' feature
  adds blank space between the left and right-hand pages
  Added -C 'crop' feature
  crops input pages according to specified values
  specified as -C 10,20,30,40
  in the order left,right,top,bottom
  Added -D 'divide pages' feature
  divides each page of the input PDF into two pages
  useful for handling PDFs of scanned books where they
  scanned two pages at a time and didn't separate them
  for you
  Added -e 'export preview JPEG' function
  exports a preview of 2 (combined) pages to check before
  running the full command
  Used as liesel -i input.pdf -e 5,6 -o output.jpeg
  Found and patched a bug in the has_ending() function
  Was previously case-sensitive, and so would reject
  input files where .pdf was written in alternate cases
  for example, as .PDF or .Pdf

 -- rail5 <andrew@rail5.org>  Sat, 25 Dec 2021 18:26:08 -0500

liesel (7.2.1) focal; urgency=low

  * Found and patched two bugs
  Page '0' was permissible input (-r function) in some cases
  and would of course crash the program
  That's now patched
  iswritable() check failed on NULL paths
  if, for example, the user is outputting to a file in the
  current working directory and therefore doesn't specify
  That's now patched as well

 -- rail5 <andrew@rail5.org>  Wed, 22 Dec 2021 04:53:33 -0500

liesel (7.2) focal; urgency=low

  * Liesel can now also be cross-compiled
  to a fully-functioning Windows binary
  Next step is to modify a thing or two in the BookThief GUI
  Pascal of course is made for cross-compilation, but the
  TProc that's supposed to spawn Liesel has given
  some trouble on Windows, and I don't understand Windows
  well enough to know why just yet

 -- rail5 <andrew@rail5.org>  Mon, 20 Dec 2021 04:32:17 -0500

liesel (7.1.2) focal; urgency=low

  * Patched that bug in the filesystem permissions checking
  Hadn't realized that the unistd access(path, W_OK) checked
  not only whether the path was writable (as it seemed to me
  from the docs) but also returned a failure if the specified
  file didn't already exist. To my mind those are two distinct

classes/liesel.cpp
#include "liesel.h"

using namespace std;

bool Liesel::Book::load_document(const string &input, bool pdfstdin, bool speak) {	
	if (!pdfstdin) {
		if (!file_exists(input.c_str())) {
			if (speak) {
			cerr << "Error: File '" << input << "' not found" << endl;
			}
			return false;
		}
		if (!has_ending(input, ".pdf")) {
			if (speak) {
				cerr << "Error: At this stage, Liesel only supports PDFs\nPlease look for a future version to support other formats" << endl;
			}
			return false;
		}
		document = poppler::document::load_from_file(input.c_str());
					
	} else {
		document = poppler::document::load_from_raw_data(input.c_str(), input.size());
	}
	loaded = true;
	return true;
}
		
void Liesel::Book::count_pages(bool verbose) {
	if (loaded) {
		pagecount = document->pages();
		if (verbose) {
			cout << "Pages counted: " << pagecount << endl;
		}
	}
}
		
bool Liesel::Book::set_pages(bool rangeflag, string rangevalue) {
	selectedpages.clear();
	
	if (rangeflag) {
		vector<string> multiranges = explode(rangevalue, ',');
		
		for (long unsigned int i=0;i<multiranges.size();i++) {
			vector<string> singlerange = explode(multiranges[i], '-');
			if (singlerange.size() > 2) {
				cerr << "Error: Invalid range '" << multiranges[i] << "'" << endl;
				return false;
			}
			
			if (singlerange.size() > 1) {
				if (!is_number(singlerange[0]) || !is_number(singlerange[1])) {
					cerr << "Error: Invalid (non-numeric) range '" << multiranges[i] << "'" << endl;
					return false;
				}
				
				int startpage = stoi(singlerange[0]);
				int endpage = stoi(singlerange[1]);
				
				if (startpage == endpage || min(startpage, endpage) == 0) {
					cerr << "Error: Invalid range '" << multiranges[i] << "'" << endl;
					return false;
				}
			
				if (max(startpage, endpage) > pagecount) {
					cerr << "Error: Given range value '" << multiranges[i] << "' out of range for supplied PDF" << endl;
					return false;
				}
				
				for (int x=startpage-1;x>=endpage-1;x--) {
					// Descending ranges (ie, -r 10-1)
					selectedpages.push_back(x);
				}
			
				for (int x=startpage-1;x<endpage;x++) {
					// Ascending (normal) ranges (ie, -r 1-10)
					selectedpages.push_back(x);
				}
			} else if (singlerange.size() == 1) {
				if (!is_number(singlerange[0])) {
					cerr << "Error: Invalid (non-numeric) range '" << multiranges[i] << "'" << endl;
					return false;
				}
				
				int singlepagerange = stoi(singlerange[0]);
				
				if (singlepagerange > pagecount) {
					cerr << "Error: Page '" << singlepagerange << "' out of range for supplied PDF" << endl;
					return false;
				}
				
				if (singlepagerange == 0) {
					cerr << "Error: Invalid page '0'" << endl;
					return false;
				}
				
				selectedpages.push_back(singlepagerange-1);
			} else {
				// User inputted two commas by mistake (-r 1-20,21-40,,,,,,44-45)
				continue;
			}
		}
	} else {
		// Printing entire PDF
		for (int i=0;i<pagecount;i++) {
			selectedpages.push_back(i);
		}
	}
	return true;
}

void Liesel::Book::calculate_segments() {
	int pgcount = selectedpages.size();
	
	if (printjob.segsize > pgcount || printjob.segsize == 0) {
		printjob.segsize = pgcount;
	}
	
	double rdsegcount = (double)pgcount/printjob.segsize;
	printjob.segcount = ceil(rdsegcount);
	
	printjob.finalsegsize = printjob.segsize;
	
	if (pgcount % printjob.segsize != 0) {
		printjob.finalsegsize = pgcount % printjob.segsize;
	}
	
	if (printjob.finalsegsize < printjob.minsize && printjob.segcount > 1) {
		printjob.segcount = printjob.segcount - 1;
		printjob.finalsegsize = (pgcount % printjob.segsize) + printjob.segsize;
	}
	
	printjob.thisseg = 1;
	printjob.startfrom = 0;
	
	/* Keep outfile sanity checks in liesel.cpp */
	
	/* Same goes for checkflag */
}

void Liesel::Book::run_job(bool verbose, bool bookthief, bool pdfstdout) {
	while (printjob.thisseg < printjob.segcount) {
		printjob.startfrom = printjob.segsize*(printjob.thisseg-1);
		
		string newname = printjob.outfilename.substr(0, printjob.outfilename.size()-4) + "-part" + to_string(printjob.thisseg) + ".pdf";
		
		printjob.endat = printjob.startfrom + printjob.segsize;
		
		load_pages(verbose, bookthief);
		make_booklet(verbose, bookthief);
		rescale(verbose, bookthief);
		
		pages.clear();
		if (!pdfstdout) {
			if (verbose) {
				cout << endl << "Writing to file..." << endl;
			}
		
			lhdoc.lhdoc_from_image_vector(booklet);
			lhdoc.write_to_file(newname);
			
			double dpercentdone = (double)printjob.thisseg/printjob.segcount;
			int percentdone = floor(dpercentdone * 100);
			
			if (bookthief) {
				cout << percentdone << "%" << endl;
			}
		} else {
			lhdoc.lhdoc_from_image_vector(booklet);
			lhdoc.write_to_stdout();
		}
			
		
		booklet.clear();
		printjob.thisseg = printjob.thisseg + 1;
	}
	
	printjob.startfrom = printjob.segsize*(printjob.thisseg-1);
	
	string appendtofname = ".pdf";
	
	if (printjob.segcount > 1) {
		appendtofname = "-part" + to_string(printjob.thisseg) + ".pdf";
	}
	
	string newname = printjob.outfilename;
	
	if (!printjob.previewonly) {
		newname = printjob.outfilename.substr(0, printjob.outfilename.size()-4) + appendtofname;
	}

	printjob.endat = printjob.startfrom + printjob.finalsegsize;
	
	load_pages(verbose, bookthief);
	make_booklet(verbose, bookthief);
	rescale(verbose, bookthief);
	
	if (!pdfstdout) {
		if (verbose) {
			cout << endl << "Writing to file..." << endl;
		}
		
		if (printjob.previewonly) {
			writeImages(booklet.begin(), booklet.end(), newname); // Use GraphicsMagick for 'export preview JPEG'
			return;
		}
		
		lhdoc.lhdoc_from_image_vector(booklet);
		lhdoc.write_to_file(newname);
		
		if (bookthief) {
			cout << "100%" << endl;
		}
		cout << "Done!" << endl;
	} else {
		lhdoc.lhdoc_from_image_vector(booklet);
		lhdoc.write_to_stdout();
	}
}

void Liesel::Book::display_progress(int progress, int stage) {
	int number_of_stages = 2; // "Rescaling" is made obsolete with this change
	int progcounter = (progress / number_of_stages) + ((stage - 1) * (100 / number_of_stages)); // Calculate initial value
	
	progcounter = (progcounter / printjob.segcount) + ((printjob.thisseg - 1) * (100 / printjob.segcount)); // Divide it, based on the number of segments
	
	bool subtractor = progcounter; // Subtract 0 from 0, 1 from 1,2,3,4,etc
	
	cout << progcounter-subtractor << "%" << endl;
}

void Liesel::Book::load_pages(bool verbose, bool bookthief) {
	if (verbose) {
		cout << "Ready to load PDF pages" << endl;
	}
	
	int vectorindex = 0;
	int y = 0;
	
	for (int i=printjob.startfrom;i<printjob.endat;i++) {
		poppler::page_renderer imagizer;
		imagizer.set_image_format(poppler::image::format_argb32); // Format set to ARGB32 so that the render can be transferred to the Magick++ Image class
		
		poppler::image loadedpage = imagizer.render_page(document->create_page(selectedpages[i]), properties.quality, properties.quality); // Render page as image
		
		Magick::Image page(loadedpage.width(), loadedpage.height(), "BGRA", Magick::StorageType::CharPixel, loadedpage.data()); // Copy raw image bytes to Magick++ for further processing
		
		page.quality(properties.quality);
		page.resolutionUnits(Magick::PixelsPerInchResolution);
		page.density(Magick::Geometry(properties.quality, properties.quality));
		page.matte(false);
		
		if (properties.grayscale) {
			page.type(Magick::GrayscaleType);
		}
		
		if (properties.alterthreshold) {
			page.threshold(properties.threshold);
		}
		
		if (properties.dividepages) {
			size_t halfwidth = page.columns() / 2;
			size_t dividedheight = page.rows();
			
			Magick::Geometry lefthalf = Magick::Geometry(halfwidth, dividedheight, 0, 0);
			Magick::Geometry righthalf = Magick::Geometry(halfwidth, dividedheight, halfwidth, 0);
			
			Magick::Image newpage = page;
			newpage.crop(lefthalf);
			page.crop(righthalf);
			
			pages.push_back(newpage); // Add the left half (newpage), page becomes the right half
		}
		
		pages.push_back(page);
		
		if (properties.cropflag) {
			for (long unsigned int x=0;x<properties.cropvalues.size();x++) {
				if (properties.cropvalues[x] > 100) {
					properties.cropvalues[x] = 100;
				}
			}
			
			int cropleft = ((pages[vectorindex].columns() / 100) * properties.cropvalues[0]) / 2;
			int cropright = ((pages[vectorindex].columns() / 100) * properties.cropvalues[1]) / 2;
			int croptop = ((pages[vectorindex].columns() / 100) * properties.cropvalues[2]) / 2;
			int cropbottom = ((pages[vectorindex].columns() / 100) * properties.cropvalues[3]) / 2;
			
			Magick::Geometry lefthand(pages[vectorindex].columns() - cropleft, pages[vectorindex].rows(), cropleft, 0);
			pages[vectorindex].crop(lefthand);
			
			Magick::Geometry righthand(pages[vectorindex].columns() - cropright, pages[vectorindex].rows(), 0, 0);
			pages[vectorindex].crop(righthand);
			
			Magick::Geometry tophand(pages[vectorindex].columns(), pages[vectorindex].rows() - croptop, 0, croptop);
			pages[vectorindex].crop(tophand);
			
			Magick::Geometry bottomhand(pages[vectorindex].columns(), pages[vectorindex].rows() - cropbottom, 0, 0);
			pages[vectorindex].crop(bottomhand);
			
			if (properties.dividepages) {
				// Crop the righthalf as well
				vectorindex = vectorindex + 1;
				
				Magick::Geometry lefthand(pages[vectorindex].columns() - cropleft, pages[vectorindex].rows(), cropleft, 0);
				pages[vectorindex].crop(lefthand);
				
				Magick::Geometry righthand(pages[vectorindex].columns() - cropright, pages[vectorindex].rows(), 0, 0);
				pages[vectorindex].crop(righthand);
				
				Magick::Geometry tophand(pages[vectorindex].columns(), pages[vectorindex].rows() - croptop, 0, croptop);
				pages[vectorindex].crop(tophand);
				
				Magick::Geometry bottomhand(pages[vectorindex].columns(), pages[vectorindex].rows() - cropbottom, 0, 0);
				pages[vectorindex].crop(bottomhand);
			}
		}
		
		vectorindex = vectorindex + 1;
		y = y + 1;
		
		if (verbose) {
			cout << "Page " << selectedpages[i]+1 << " loaded... " << flush;
		}
		
		if (bookthief) {
			double dprog = (double)(i + 1)/(printjob.endat + 1)*100;
			int prog = floor(dprog);
			display_progress(prog, 1);
		}
	}
	
	size_t width = pages[0].columns();
	size_t height = pages[0].rows();
	
	Magick::Geometry blanksize = Magick::Geometry(width, height);
	
	Magick::Color blankcolor("white");
	
	Magick::Image blank_image(blanksize, blankcolor);
	
	int current_total = pages.size();
	int number_of_blanks = 0;
	
	if (current_total % 2 != 0 && !printjob.previewonly) {
		number_of_blanks = number_of_blanks + 1;
		current_total = current_total + 1;
		if (verbose) {
			cout << endl << "Blank page added to make an even number" << endl;
		}
	}
	
	if (current_total % 4 != 0 && !printjob.previewonly) {
		number_of_blanks = number_of_blanks + 2;
		// No further need to update current_total
		if (verbose) {
			cout << endl << "Two blank pages added to make divisible by 4" << endl;
		}
	}
	
	for (int appending=0;appending<number_of_blanks;appending++) {
		pages.push_back(blank_image);
	}

	if (verbose) {
		cout << endl << "All pages loaded" << endl;
	}
}

void Liesel::Book::make_booklet(bool verbose, bool bookthief) {

	if (printjob.linear_output) {
		booklet = move(pages);
		return;
	}

	int relevantpagecount = pages.size();
	int relevantpagecountfromzero = relevantpagecount - 1;
	
	int first = 0;
	int second = relevantpagecountfromzero;
	
	int minmargin = properties.widenby;
	
	bool local_widenflag = properties.widenflag;
	int local_widenby = properties.widenby;
	
	Magick::Geometry blanksize = Magick::Geometry(50, 50);
	Magick::Color blankcolor("white");
	
	while (first <= (relevantpagecountfromzero / 2)) {
		Magick::Image newimg(blanksize, blankcolor);
		newimg.resolutionUnits(Magick::PixelsPerInchResolution);
		newimg.density(Magick::Geometry(properties.quality, properties.quality));
		
		size_t widthone = pages[pages.size()-1].columns();
		size_t heightone = pages[pages.size()-1].rows(); // Width and height of the final page in the pages vector
		
		size_t widthtwo = pages[0].columns();
		size_t heighttwo = pages[0].rows(); // Width and height of the first page in the pages vector
		
		size_t width = max(widthone, widthtwo) * 2; // Pick the wider of the two, and double that space
		size_t height = max(heightone, heighttwo); // Pick the higher of the two heights
		
		if (heightone != heighttwo || widthone != widthtwo) {
			Magick::Geometry matchsize = Magick::Geometry(width/2, height);
			matchsize.aspect(true);
			pages[pages.size()-1].resize(matchsize);
			pages[0].resize(matchsize); // Rescale them to be equal sizes
		}
		
		if (properties.automargin) {
			local_widenflag = true;
			if (properties.maxmargin != 0) {
				double widenconst = (double)properties.maxmargin / (double)(relevantpagecountfromzero / 2);
				local_widenby = first * widenconst;
			} else {
				local_widenby = first * 0.35;
			}
			
			if (local_widenby < minmargin) {
				local_widenby = minmargin;
			}
		}
		
		double dwidenby = ((width / 4) / 100) * local_widenby;
		
		if (local_widenflag) {
			size_t widthwithmargin = (width / 2) - dwidenby;
			Magick::Geometry resizedagain = Magick::Geometry(widthwithmargin, height);
			resizedagain.aspect(true);
			
			pages[pages.size()-1].resize(resizedagain);
			pages[0].resize(resizedagain);
		}
		
		Magick::Geometry newsize = Magick::Geometry(width, height);
		newsize.aspect(true);
		
		newimg.resize(newsize);
		
		if (properties.dividepages && printjob.previewonly) {
			newimg.composite(pages[pages.size()-1], ((width / 2) + (dwidenby * local_widenflag)), 0);
				/*
				Same old trick of using a boolean in math
				If we're not widening, then this goes at position: width/2 + (some amount of widening * 0)
					ie, width/2
				Otherwise,
					width/2 + (some amount of widening * 1)
				*/
			newimg.composite(pages[0], 0, 0);
		} else {
			newimg.composite(pages[pages.size()-1], 0, 0);
			newimg.composite(pages[0], ((width / 2) + (dwidenby * local_widenflag)), 0);
		}
		
		if (!printjob.previewonly || printjob.rescaling) {
			newimg.rotate(90);
		}
		
		booklet.push_back(newimg);
		
		if (verbose) {
			cout << "Combined pages " << second+1 << " and " << first+1 << "... " << flush;
		}
		
		if (bookthief) {
			double dprog = (double)(first + 1)/((relevantpagecountfromzero / 2) + 1)*100;
			int prog = floor(dprog);
			display_progress(prog, 2);
		}
		
		newimg.rotate(-90);
		
		pages.pop_back(); // Delete the last page from the pages vector
		pages.erase(pages.begin()); // Delete the first page from the pages vector
		
		first = first + 1;
		second = second - 1;
		
		if (first <= (relevantpagecountfromzero / 2)) {
			newimg = Magick::Image(blanksize, blankcolor);
			
			newimg.resolutionUnits(Magick::PixelsPerInchResolution);
			newimg.density(Magick::Geometry(properties.quality, properties.quality));
			
			widthone = pages[pages.size()-1].columns();
			heightone = pages[pages.size()-1].rows(); // Width and height of the final page in the pages vector
			
			widthtwo = pages[0].columns();
			heighttwo = pages[0].rows(); // Width and height of the first page in the pages vector
			
			width = max(widthone, widthtwo) * 2; // Pick the wider of the two, and double that space
			height = max(heightone, heighttwo); // Pick the higher of the two heights
			
			if (heightone != heighttwo || widthone != widthtwo) {
				Magick::Geometry matchsize = Magick::Geometry(width/2, height);
				matchsize.aspect(true);
				pages[pages.size()-1].resize(matchsize);
				pages[0].resize(matchsize); // Rescale them to be equal sizes
			}
			
			dwidenby = ((width / 4) / 100) * local_widenby;
			
			if (local_widenflag) {
				size_t widthwithmargin = (width / 2) - dwidenby;
				Magick::Geometry resizedagain = Magick::Geometry(widthwithmargin, height);
				resizedagain.aspect(true);
				pages[pages.size()-1].resize(resizedagain);
				pages[0].resize(resizedagain);
			}
			
			newsize = Magick::Geometry(width, height);
			newsize.aspect(true);
			
			newimg.resize(newsize);
			
			newimg.composite(pages[0], 0, 0);
			newimg.composite(pages[pages.size()-1], ((width / 2) + (dwidenby * local_widenflag)), 0);
			
			if (!printjob.landscapeflip) {
				newimg.rotate(90);
			} else {
				newimg.rotate(-90);
			}
			
			booklet.push_back(newimg);
			
			if (verbose) {
				cout << "Combined pages " << first+1 << " and " << second+1 << "... " << flush;
			}
			
			if (bookthief) {
				double dprog = (double)(first + 1)/((relevantpagecountfromzero / 2) + 1)*100;
				int prog = floor(dprog);
				display_progress(prog, 2);
			}
			
			newimg.rotate(-90);
			
			pages.pop_back();
			pages.erase(pages.begin());
			
			first = first + 1;
			second = second - 1;
		}
		
	}
	
	if (verbose) {
		cout << endl << "New PDF prepared. Ready to write to file..." << endl;
	}
}

void Liesel::Book::rescale(bool verbose, bool bookthief) {
	size_t width = booklet[0].columns();
	size_t height = booklet[0].rows();
	
	width = (properties.quality / 72) * width;
	height = (properties.quality / 72) * height;
	
	if (printjob.rescaling) {
		width = 72*printjob.rescale_width;
		height = 72*printjob.rescale_height;
	}
	
	lhdoc.width = width;
	lhdoc.height = height;
	
	if (printjob.previewonly) {
		if (printjob.rescaling) {
			Magick::Geometry magick_rescaled(100*printjob.rescale_width, 100*printjob.rescale_height);
			magick_rescaled.aspect(true);
			booklet[0].resize(magick_rescaled);
			if (!printjob.linear_output) {
				booklet[0].rotate(-90);
			}
		}
	}
	
}

Liesel::Book Liesel::load_book(const string &input, bool pdfstdin, bool speak) {
	Book newbook;
	newbook.load_document(input, pdfstdin, speak);
	return newbook;
}
